CASSANDRA-18071 trunk: Allow to use user-defined functions (UDF) as masking functions

adelapena opened the pull request:
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

adelapena committed:
Add masking functions to column metadata

Allows to attach the native masking functions added by CASSANDRA-17941 to the definitions of
table columns in the schema, as defined by CEP-20. The functions masking a column can be
specified on CREATE TABLE queries, right at the end of the column definition. The mask of a
column can also be changed or dropped with an ALTER TABLE query. Once a column is masked,
SELECT queries will always return the masked value of the column. That masking is done on the
coordinator, at the end of the query execution. Thus, masking won't affect any filters or
ordering, which would be based on the clear values of the masked columns. Column masks are
stored on the system table system_schema.column_masks.

patch by Andrés de la Peña; reviewed by Benjamin Lerer and Berenguer Blasi for CASSANDRA-18068

bereng commented on src/java/org/apache/cassandra/cql3/statements/schema/DropFunctionStatement.java:
```diff
@@ -120,9 +121,28 @@ public Keyspaces apply(Keyspaces schema)
+                                               .collect(joining(", "));
+
+        if (!dependentMasks.isEmpty())
+            throw ire("Function '%s' is still referenced by column masks in tables %s", name, dependentMasks);
```
I would add the keyspace name here "ks.table". Countless times I've amended error messages just bc a user had the same exact ks but: ks_dev, ks_testWhatever, ks_preProd, etc

Yes it's immediately obvious to the user. But on captured output, logs, etc... I would still add it if easy enough.

adelapena replied:
The `name` var is the qualified name of the function, so it already contains the name of the keyspace. That `name` var is created a few lines above, at the beginning of the method.

bereng replied:
Wfm Thx

bereng commented on test/distributed/org/apache/cassandra/distributed/test/ColumnMaskTest.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    private static void withAuthenticatedSession(IInvokableInstance instance, String username, String password, Consumer<Session> consumer)
+    {
```
Is this the code we would need [here](https://github.com/apache/cassandra/pull/2125#pullrequestreview-1296963803)? :thinking:

adelapena replied:
Not only. We would need something similar for authentication, but we can't use `callOnInstance` on upgrade tests. Also, as it's mentioned on that conversation we would also need a custom load balancing policy since the upgrade test involves multiple nodes. And, since upgrade tests are more sensitive to OOMs and timeouts, we would need to split the test. I have given it a go on top of 18069, [on this commit](https://github.com/apache/cassandra/pull/2125/commits/58e4cb9f98c462baa80ca1b6be43ccc376b34f7f). Some repeated runs can be found [here](https://app.circleci.com/pipelines/github/adelapena/cassandra/2650/workflows/a75e1832-bd9e-4eba-947b-e7782d6457d2).

bereng commented on test/unit/org/apache/cassandra/cql3/functions/masking/ColumnMaskWithUDFTest.java:
```diff
@@ -0,0 +1,284 @@
+    @SuppressWarnings("resource")
+    public void testUDF() throws Throwable
+    {
+        // create a table masked with and UDF and with a materialized view
```
and -> a

adelapena replied:
Fixed, thanks

adelapena committed:
Add UNMASK permission

The new permission allows to see the clear data of columns with an attached mask.
Superusers have it by default, whereas regular users don't have it.

patch by Andrés de la Peña; reviewed by Benjamin Lerer and Berenguer Blasi for CASSANDRA-18069

adelapena committed:
Add SELECT_MASKED permission

The new permission allows to run SELECT queries restricting the clear values of masked columns.
Superusers have it by default, whereas regular users don't have it.

patch by Andrés de la Peña; reviewed by Benjamin Lerer and Berenguer Blasi for CASSANDRA-18070

adelapena committed:
Add support for using UDFs as masking functions

patch by Andrés de la Peña; reviewed by Benjamin Lerer and Berenguer Blasi for CASSANDRA-18071

adelapena committed:
DO NOT MERGE - CircleCI config

blerer commented on src/java/org/apache/cassandra/cql3/statements/schema/DropFunctionStatement.java:
```diff
@@ -120,9 +121,28 @@ public Keyspaces apply(Keyspaces schema)
         if (!dependentAggregates.isEmpty())
             throw ire("Function '%s' is still referenced by aggregates %s", name, dependentAggregates);
 
+        String dependentMasks = keyspace.tables.stream()
+                                               .filter(table -> hasDependingMask(table, function))
+                                               .map(table -> table.name)
+                                               .collect(joining(", "));
```
The approach that is mostly use within the code base is to delegate that type of call to the classes themselves. See `dependentAggregates` bellow.

adelapena replied:
I have just tried to do that with two new `KeyspaceMetadata#tablesUsingFunction(Function)` and `TableMetadata#dependsOn(Function)` methods.

blerer reviewed (commented):
Could you add a test that use an UDF from another keyspace?

adelapena committed:
Address review comments

adelapena commented:
> Could you add a test that use an UDF from another keyspace?

That's already tested [here](https://github.com/apache/cassandra/blob/c41e567fbfddff8660e07109a0c6c7ada547465b/test/unit/org/apache/cassandra/cql3/functions/masking/ColumnMaskWithUDFTest.java#L210-L221).

pmcfadin commented:
Closing as [CASSANDRA-18071](https://issues.apache.org/jira/browse/CASSANDRA-18071) has been resolved as Fixed — if this PR contains changes not captured in the merged fix, please reopen.
